### ISSUE TITLE
cmscp: report to dashboard the long exit code

### ISSUE THREAD

**Issue by mmascher** [dmwm/CRABServer#4642]
The "application by exit code" chart in Dashboard (see http://dashb-cms-jobsmry.cern.ch/dashboard/request.py/dailysummary#button=successfailures&sites%5B%5D=All+T3210&submissions%5B%5D=crab3&sitesSort=7&start=null&end=null&timerange=lastWeek&granularity=Hourly&generic=0&sortby=1&series=All) reports the short exito code for the jobs (e.g. 65 instead of 8001)

It seems to me that the problem is that cmscp uses the value coming from the command line here https://github.com/dmwm/CRABServer/blob/master/scripts/gWMS-CMSRunAnalysis.sh#L136 instead of the one the job wrappers saves in the fwjr.

In general I believe we should take the exit code og the job from the fwjr and not from the command line: https://github.com/dmwm/CRABServer/blob/master/scripts/cmscp.py#L1840-L1845
The only exception is probably if there is no fwjr available for some reason (I believe it can happen). In this case I would keep the current code.

**Comment by belforte** [comment]
I disagree.

On 01/26/2015 10:11 AM, Marco Mascheroni wrote:
> The only exception is probably if there is no fwjr available for some reason (I believe it can
> happen). In this case I would keep the current code.

fwjr can miss exit code for two reasons:
1. application (e.g. cmsRun) failed to produce it while it was supposed to, so
   a specific (long) exit code needs to be set
2. application (e.g. user's script exe, details to be defined) is not expected
   to produce it in the first place. In this case we still prefer to group all
   cases where application failed into a single category (for easier overview)
   rather then disperse in all possible posix exit codes like Crab2 was doing
   in the past.

Exit code is there mostly to allow us to classify failed jobs into categories
due to common source, so that we can decide where to direct attention
based on how many and how much resources are wasted there.
A secondary goal is to point users in the right direction in case this
is something they can fix. For that "your custom script ended with error
and this message... reporting the last lines of their stdout" may be enough.

**Comment by mmascher** [comment]
Hi Stefano,

I should have mentioned that I am talking about the job wrapper json fjr, which is hardly missing. In both the scriptExe and cmsRun cases the jobReport.json is written by the wrapper if a FrameworkJobReport.xml is not produced by the user executable.

And this jobReport.json is always supposed to be written by the job wrapper (CMSRunAnalysis.py).

So, I believe that reporting the short exit code in cmscp would still be useful since it would give us more information about what went wrong and why a long exit code was not produced. And I can only see three cases why this is happening:

1) the job wrapper is killed (there is no signal handler)
2) we can't write the jobReport.json in the job wrapper file for some reason (even though a permission error whould be spotted before, but who knows)
3) there is a "bug" in the job wrapper and we exit before writing the jobReport.json

Reporting the short exit code would allow us to distiguish between 1 and 2/3 imho

**Comment by belforte** [comment]
can you summarize in one line relation between job wrapper and cmscp ? How does it happens that job wrapper is killed and cmscp is not ? who talks to dashboard ? And what if there is a bug in cmscp ?

I still think you want an exit code that means "step X in the wrapper/cmsp chaing failed", not
"something failed with exit code Y and we think this usually means step N since bugs aside that's the only time where we pass on the exit code from some scripts rather then setting explicitely one from the table"
I suspect like would be easier if unknown exit code always means 3) in your list.

**Comment by mmascher** [comment]
On Mon, Jan 26, 2015 at 12:03 PM, Stefano Belforte
<notifications@github.com> wrote:
>
> can you summarize in one line relation between job wrapper and cmscp ? How does it happens that job wrapper is killed and cmscp is not ? who talks to dashboard ? And what if there is a bug in cmscp ?

There is a script (which probably should also be called jobwrapper :))
that executes both what I was calling jobwrapper before, and then
cmscp.

>
> I still think you want an exit code that means "step X in the wrapper/cmsp chaing failed", not
> "something failed with exit code Y and we think this usually means step N since bugs aside that's the only time where we pass on the exit code from some scripts rather then setting explicitely one from the table"
> I suspect like would be easier if unknown exit code always means 3) in your list.

Well, I guess we can go with one exit code which is the user's
couterpart of: "cmscp cannot find jobReport.json" and then remove the
8bit exit code completely.

For reference right now we are reporting this when the first part of
the job wrapper finishes:

Dashboard end parameters: {'MonitorID':
'150121_215515_crab3test-1:jlee_crab_20150121_225443', 'ExeExitCode':
8001, 'JobExitCode': 8001, 'MonitorJobID':
'1_https://glidein.cern.ch/1/150121:215515:crab3test-1:jlee:crab:20150121:225443_0'}

and then this when cmscp finishes:

Dashboard stageout failure parameters: {'MonitorID':
'150121_215515_crab3test-1:jlee_crab_20150121_225443', 'MonitorJobID':
'1_https://glidein.cern.ch/1/150121:215515:crab3test-1:jlee:crab:20150121:225443_0',
'JobExitCode': 65}

**Comment by belforte** [comment]
in this case cmscp should simply report 8001.
If you job wrapper failed to produce jobReport.json, or
produces one w/o the Exit code information I
would start with:
  80000 - Internal error in CRAB job wrapper
and see how many of those we get before worrying about
dividing the category further.

I would leave to job wrapper to handle possibly
known/standard failure of user's exe (segfault=11 e.g.)

It may help at some point to mark in the twiki
https://twiki.cern.ch/twiki/bin/view/CMSPublic/JobExitCodes
which ones Crab3 uses and which one were only for Crab2 and
we can remove. That list is way too long.
